sound/soc/codecs/rt711.c

```diff
@@ -426,19 +413,39 @@ static int rt711_set_jack_detect(struct snd_soc_component *component,
 	rt711_index_write(rt711->regmap, RT711_VENDOR_REG,
 		0x19, 0x2e11);
 
-	/* power off */
-	if (dapm->bias_level <= SND_SOC_BIAS_STANDBY)
-		regmap_write(rt711->regmap,
-			RT711_SET_AUDIO_POWER_STATE, AC_PWRST_D3);
 	mutex_unlock(&rt711->calibrate_mutex);
 
+	mod_delayed_work(system_power_efficient_wq,
+			   &rt711->jack_detect_work, msecs_to_jiffies(250));
+
+}
+
+static int rt711_set_jack_detect(struct snd_soc_component *component,
+	struct snd_soc_jack *hs_jack, void *data)
+{
+	struct rt711_priv *rt711 = snd_soc_component_get_drvdata(component);
+	struct snd_soc_dapm_context *dapm =
+		snd_soc_component_get_dapm(component);
+
 	rt711->hs_jack = hs_jack;
 
+	if (!rt711->hw_init) {
+		dev_warn(&rt711->slave->dev, "%s hw_init not ready yet\n", __func__);
+		return 0;
+	}
+
 	dev_dbg(&rt711->slave->dev, "in %s...\n", __func__);
+	/* power on */
+	if (dapm->bias_level <= SND_SOC_BIAS_STANDBY)
+		regmap_write(rt711->regmap,
+			RT711_SET_AUDIO_POWER_STATE, AC_PWRST_D0);
 
-	mod_delayed_work(system_power_efficient_wq,
-			   &rt711->jack_detect_work, msecs_to_jiffies(250));
+	rt711_jack_init(rt711);
 
+	/* power off */
+	if (dapm->bias_level <= SND_SOC_BIAS_STANDBY)
+		regmap_write(rt711->regmap,
+			RT711_SET_AUDIO_POWER_STATE, AC_PWRST_D3);
```

Comment:
@bardliao Could you move the power on/off to rt711_jack_init function?
It has a chance to be power off state when running rt711_jack_init function.